rfc/rfc-63/rfc-63.md

```diff
@@ -0,0 +1,418 @@
+|------------------------------|----------------------|----------------|----------------|-------------------------|
+| org_id=1/datestr=2022-10-01/ | base_file_1.parquet  | `2022-10-01`   | `2022-10-01`   | Old partitioning scheme |
+| org_id=1/datestr=2022-10-02/ | base_file_2.parquet  | `2022-10-02`   | `2022-10-02`   |                         |
+| org_id=2/datestr=2022-10-01/ | base_file_3.parquet  | `2022-10-01`   | `2022-10-01`   |                         |
+| org_id=3/datestr=2022-10-01/ | base_file_4.parquet  | `2022-10-01`   | `2022-10-01`   |                         |
+| ...                          | ...                  | ...            | ...            | ...                     |
+| org_id=1/                    | base_file_10.parquet | `2022-10-10`   | `2022-10-11`   | New partitioning scheme |
+| org_id=2/                    | base_file_11.parquet | `2022-10-10`   | `2022-10-15`   |                         |
+| ...                          | ...                  | ...            | ...            | ...                     |
+
+For the example above, even with the mix of old and new partitioning scheme, we should be able to effectively skip data
+based on the range of `datestr` for each file, regardless how the files are stored under different physical partition
+paths in the table.
+
+### Use Case 3: Support for different indexes
+
+Functional index framework should be able to work with different index types such as bloom index, column stats, and at
+the same time should be extensible enough to support any other secondary index such
+as [vector](https://www.pinecone.io/learn/vector-database/) [index]((https://weaviate.io/developers/weaviate/concepts/vector-index))
+in the future. If we think about a very simple index on a column, it is kind of a functional index with identity
+function `f(X) = X`. It is important to note that these are secondary indexes in the sense they will be stored
+separately from the data, and not materialized with the data.
+
+## Goals and Non-Goals
+
+Based on the use cases we plan to support, we set the following goals and non-goals for this RFC.
+
+### Goals
+
+- Modular, easy-to-use indexing subsystem, with full SQL support to manage indexes.
+- Absorb partitioning into indexes and aggregate statistics at the storage partition level. 
+- Support efficient data skipping with different indexing mechanisms.
+- Be engine-agnostic and language-agnostic.
+
+### Non-Goals
+
+- DO NOT remove physical partitioning, which remains as an option for physically storing data in different folders and
+  partition pruning. Viewing partitions as yet another index goes beyond the traditional view as pointed in use case 2,
+  and we will see how we can support logical partitioning and partition evolution simply with indexes.
+- DO NOT tackle the support of using these indexes on the write path in this RFC. That said, we will present a glimpse
```

Comment:
this can naturally happen, as we reimplement existing indexes as a read on the multi modal index with hints?

Reply:
By hints, do you mean something like `INSERT INTO tbl /*+ hint [ , ... ] */ VALUES ...`? If so, I think that's a great idea. We can think of introducing hints in general (independent of functional index) even for other SELECT queries.

Reply:
Added in appendix section.